File: pattern-wise/code-solution/src/main/python/first_miss_positive.py
"""
Leetcode Problem 41: First Missing Positive

Approach:
1. The problem is solved using an in-place swapping technique.
2. Iterate through the array and place each number `nums[i]` at its correct index `nums[i] - 1` if it is within the range [1, n].
3. After rearranging, iterate through the array again to find the first index `i` where `nums[i] != i + 1`. The missing positive is `i + 1`.
4. If all numbers are in their correct positions, the missing positive is `n + 1`.

Time Complexity: O(n)
- The array is traversed twice, making the time complexity linear.

Space Complexity: O(1)
- The rearrangement is done in-place without using extra space.

"""

def first_missing_positive(nums):
    n = len(nums)
    
    # Step 1: Place each number in its correct position
    for i in range(n):
        while 1 <= nums[i] <= n and nums[nums[i] - 1] != nums[i]:
            nums[nums[i] - 1], nums[i] = nums[i], nums[nums[i] - 1]  # Swap
    
    # Step 2: Find the first missing positive
    for i in range(n):
        if nums[i] != i + 1:
            return i + 1
    
    return n + 1

def main():
    nums = [3, 4, -1, 1] # Output: 2
    print(f"First Missing Positive in {nums} is {first_missing_positive(nums)}")

    nums = [1, 2, 0] # Output: 3
    print(f"First Missing Positive in {nums} is {first_missing_positive(nums)}")

    nums = [7, 8, 9, 11, 12] # Output: 1
    print(f"First Missing Positive in {nums} is {first_missing_positive(nums)}")

    nums = [1] # Output: 2
    print(f"First Missing Positive in {nums} is {first_missing_positive(nums)}")

if __name__ == "__main__":
    main()

File: pattern-wise/code-solution/src/main/java/FirstMissPositive.java
import java.util.Arrays;

public class FirstMissPositive {

    public static int firstMissingPositive(int[] nums) {
        int result = 1, n = nums.length, temp;

        for (int i = 0; i < n; i++) {
            while (nums[i] > 0 && nums[i] < n + 1 && nums[nums[i] - 1] != nums[i]) {
                temp = nums[i];
                nums[i] = nums[nums[i] - 1];
                nums[nums[i] - 1] = temp;
            }
        }

        for (int i = 0; i < n; i++) {
            if (i + 1 != nums[i])
                return i + 1;
        }

        return n + 1;
    }

    public static void main(String[] args) {
        int[] nums1 = {1, 2, 0};
        int[] nums2 = {3, 4, -1, 1};
        int[] nums3 = {7, 8, 9, 11, 12};

        System.out.println("First Missing Positive in " + Arrays.toString(nums1) + " is " + firstMissingPositive(nums1));
        System.out.println("First Missing Positive in " + Arrays.toString(nums2) + " is " + firstMissingPositive(nums2));
        System.out.println("First Missing Positive in " + Arrays.toString(nums3) + " is " + firstMissingPositive(nums3));
    }
}
